src/algorithms/calorimetry/ImagingTopoCluster.h

```diff
@@ -198,35 +239,57 @@ namespace eicrecon {
     }
 
     // grouping function with Breadth-First Search
-    void bfs_group(const edm4eic::CalorimeterHitCollection &hits, std::set<std::size_t> &group, std::size_t idx, std::vector<bool> &visits) const {
-      visits[idx] = true;
+    template<typename Compare>
+    void bfs_group(const edm4eic::CalorimeterHitCollection &hits, std::set<std::size_t,Compare>& indices, std::vector<std::size_t> &group, const std::size_t idx) const {
 
       // not a qualified hit to participate clustering, stop here
       if (hits[idx].getEnergy() < m_cfg.minClusterHitEdep) {
         return;
       }
 
-      group.insert(idx);
-      size_t prev_size = 0;
-
+      std::size_t prev_size = 0;
       while (prev_size != group.size()) {
-        prev_size = group.size();
-        for (std::size_t idx1 : group) {
-          // check neighbours
-          for (std::size_t idx2 = 0; idx2 < hits.size(); ++idx2) {
-            // not a qualified hit to participate clustering, skip
-            if (hits[idx2].getEnergy() < m_cfg.minClusterHitEdep) {
+        auto prev_group = group;
+        for (auto idx1 = std::next(prev_group.begin(), prev_size); idx1 != prev_group.end(); ++idx1) {
+          // check neighbours (note comments on loop over set above)
+          for (auto idx2 = indices.begin(); idx2 != indices.end();
+              indices.empty() ? idx2 = indices.end() : idx2) {
+
+            // skip idx1 and original idx
```

Comment:
```suggestion
            // skip idx1 and original idx
            // (we cannot erase idx since it would invalidate iterator in calling scope)
```